packages/expo/ios/AppDelegates/ExpoAppDelegate.swift

```diff
@@ -1,32 +1,129 @@
+
+ TODO vonovak check macOS support once RN macOS 78 goes out
  */
 @objc(EXExpoAppDelegate)
-open class ExpoAppDelegate: ExpoAppInstance {
-  /**
-   Whether to skip calling the React Native instance setup from `RCTAppDelegate`.
-   Set this property to `false` if your app delegate is not supposed to initialize a React Native app,
-   but only to handle the app delegate subscribers.
-   */
+open class ExpoAppDelegate: ExpoReactNativeFactoryDelegate, UIApplicationDelegate, UISceneDelegate {
+  /// The window object, used to render the UIViewControllers
+  public var window: UIWindow?
+
+  /// From RCTAppDelegate
+  @objc public var moduleName: String = ""
+  @objc public var initialProps: [AnyHashable: Any]?
+
+  @objc public let reactDelegate = ExpoReactDelegate(
+    handlers: ExpoAppDelegateSubscriberRepository.reactDelegateHandlers
+  )
+
+  /// If `automaticallyLoadReactNativeWindow` is set to `true`, the React Native window will be loaded automatically.
+  private var automaticallyLoadReactNativeWindow = true
+
+  // Default initializer
+  public override init() {
+    super.init()
+    self.reactNativeFactory = ExpoReactNativeFactory(delegate: self, reactDelegate: self.reactDelegate)
+  }
+
+  func loadReactNativeWindow(launchOptions: [UIApplication.LaunchOptionsKey: Any]?) {
+    guard let reactNativeFactory = self.reactNativeFactory else {
+      fatalError("loadReactNativeWindow: Missing reactNativeFactory in ExpoAppInstance")
+    }
+    let rootView = reactNativeFactory.rootViewFactory.view(
+      withModuleName: self.moduleName as String,
+      initialProperties: self.initialProps,
+      launchOptions: launchOptions
+    )
+
+    let window = UIWindow(frame: UIScreen.main.bounds)
+    let rootViewController = createRootViewController()
+
+    setRootView(rootView, toRootViewController: rootViewController)
+
+    window.windowScene?.delegate = self
+    window.rootViewController = rootViewController
+    window.makeKeyAndVisible()
+
+    self.window = window
+  }
+
   @objc
-  public var shouldCallReactNativeSetup: Bool = true
+  open override func createRootViewController() -> UIViewController {
+    return reactDelegate.createRootViewController()
+  }
+
+  @objc public override func recreateRootView(
+    withBundleURL: URL?,
+    moduleName: String?,
+    initialProps: [AnyHashable: Any]?,
+    launchOptions: [AnyHashable: Any]?
+  ) -> UIView {
+    guard let reactNativeFactory = self.reactNativeFactory else {
+      fatalError("recreateRootView: Missing reactNativeFactory in ExpoAppInstance")
+    }
+
+    let rootViewFactory = reactNativeFactory.rootViewFactory
+
+    if self.newArchEnabled() {
+      // chrfalch: rootViewFactory.reactHost is not available here in swift due to the underlying RCTHost type of the property. (todo: check)
+      assert(rootViewFactory.value(forKey: "reactHost") == nil, "recreateRootViewWithBundleURL: does not support when react instance is created")
```

Comment:
maybe we can just use `rootViewFactory.reactHost` now that it exposes the property already

Reply:
figured it out! because in **RCTRootViewFactory.h**, the RCTHost is a forwarded class. and from swift we cannot import React-RuntimeApple (where RCTHost defined).
the only solution what i can tell is adding a helper in objcpp. given we only check whether the property is nil, let's keep it as-is. let's add a backlog to do some refactoring in the future.